### PR TITLE
Sfd5311/update doc tests

### PR DESCRIPTION
**Pull request by SFD5311**
Changes to the doc tests for greater independence from index version and underlying data.  Primarily using modality -> gene queries to identify genes that align with genomic_modality parameters supplied to queries and relaxing p_values in differential expression based queries.

### CODE REVIEW

**Comment by SFD5311** [comment]
Going to take a look at server logs and expand the load tests I've been working with and see if I can recreate these test failures.  Will mark as ready for review once that's done.

File: examples/bounds.md [hunk 1]
```diff
@@ -16,6 +16,11 @@ dict_keys(['minimum_value', 'maximum_value'])
 >>> max_atac_value = atac_bounds['maximum_value']
 >>> max_codex_value = codex_bounds['maximum_value']
 
->>> assert max_codex_value > max_rna_value and max_rna_value > max_atac_value
+>>> min_rna_value = rna_bounds['minimum_value']
+>>> min_atac_value = atac_bounds['minimum_value']
+>>> min_codex_value = codex_bounds['minimum_value']
+
+
+>>> assert max_codex_value > min_codex_value and max_rna_value > min_rna_value and max_atac_value > min_atac_value
```
Comment (mccalluc) [inline examples/bounds.md]:
```suggestion
>>> assert max_codex_value > min_codex_value
>>> assert max_rna_value > min_rna_value
>>> assert max_atac_value > min_atac_value
```
So if one of them fails it's easy to tell which one.

File: examples/differential-expression.md [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
 >>> from hubmap_api_py_client import Client
 >>> client = Client(environ['API_ENDPOINT'])
 
->>> organ_name = client.select_organs().get_list()[1]['grouping_name']
->>> organ_genes = client.select_genes(where='organ', has=[organ_name], genomic_modality='rna', p_value=0.05)
->>> organ_genes_details = organ_genes.get_list()[0:10]
->>> organ_genes_details[0].keys()
-dict_keys(['gene_symbol', 'go_terms'])
+>>> gene_symbol = client.select_genes(where="modality", has=["atac"]).get_list()[10]['gene_symbol']
+>>> organs_with_gene = client.select_organs(where='gene', has=[gene_symbol], genomic_modality='atac', p_value=1.0)
```
Comment (mccalluc) [inline examples/differential-expression.md]:
```suggestion
>>> modality = 'atac'
>>> gene_symbol = client.select_genes(where='modality', has=[modality]).get_list()[10]['gene_symbol']
>>> organs_with_gene = client.select_organs(where='gene', has=[gene_symbol], genomic_modality=modality, p_value=1.0)
```
Pull out repeated string as variable.

File: examples/repr.md [hunk 3]
```diff
@@ -7,9 +7,9 @@
 >>> from hubmap_api_py_client import Client
 >>> client = Client(environ['API_ENDPOINT'])
 >>> client
```
Comment (mccalluc) [inline examples/repr.md]: I'd suggest getting rid of this test, so that you will use this test against devel enviroments, and never again have to dig out of this hole.

File: examples/differential-expression.md (hunk 2 above)
Comment (mccalluc) [reply examples/differential-expression.md]: (same advice applies below.)

**Review by mccalluc — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -66,7 +66,7 @@ Only some types of objects can be retrieved from other types of objects:
...
 | [`select_clusters()`](https://github.com/hubmapconsortium/hubmap-api-py-client/blob/main/examples/select_clusters.md)                                                                                                              | ✓         |           | ✓         | ✓         | ✓ ✩       | ✩         | ✩         |           |
 | [`select_datasets()`](https://github.com/hubmapconsortium/hubmap-api-py-client/blob/main/examples/select_datasets.md)| ✓         | ✓         | ✓         | ✓         |           |✶          |           | ✓         |
-| [`select_genes()`](https://github.com/hubmapconsortium/hubmap-api-py-client/blob/main/examples/select_genes.md)                                                                                                              | ✓         |           | ✓ ✩       |           | ✓         | ✓ ✩       | ✩         |           |
+| [`select_genes()`](https://github.com/hubmapconsortium/hubmap-api-py-client/blob/main/examples/select_genes.md)                                                                                                              | ✓         |           | ✓ ✩       |           | ✓         | ✓ ✩       | ✩         |           |
```
Comment (mccalluc) [inline README.md]:
```suggestion
| [`select_genes()`](https://github.com/hubmapconsortium/hubmap-api-py-client/blob/main/examples/select_genes.md)                                                                                                              | ✓         |           | ✓ ✩       |           | ✓         | ✓ ✩       | ✩         |           |
```

File: README.md [hunk 5]
```diff
@@ -25,13 +25,13 @@ $ export API_ENDPOINT='https://cells.dev.hubmapconsortium.org/api/'
 >>> [m for m in dir(client) if m.startswith('select_')]
 ['select_cells', 'select_clusters', 'select_datasets', 'select_genes', 'select_organs', 'select_proteins']
 
->>> gene_symbol = client.select_genes().get_list()[0]['gene_symbol']
+>>> gene_symbol = client.select_genes(where="modality", has=["rna"]).get_list()[0]['gene_symbol']
```
Comment (mccalluc) [inline README.md:26]: Can you explain the behavior change? Does the underlying API now require constraints? or is the shape of response records not consistent? or something else?

File: examples/differential-expression.md [hunk 6]
```diff
@@ -1,23 +1,26 @@
...
 >>> from hubmap_api_py_client import Client
 >>> client = Client(environ['API_ENDPOINT'])
 
->>> organ_name = client.select_organs().get_list()[1]['grouping_name']
->>> organ_genes = client.select_genes(where='organ', has=[organ_name], genomic_modality='rna', p_value=0.05)
->>> organ_genes_details = organ_genes.get_list()[0:10]
->>> organ_genes_details[0].keys()
-dict_keys(['gene_symbol', 'go_terms'])
+>>> modality = 'atac'
+>>> gene_symbol = client.select_genes(where='modality', has=[modality]).get_list()[10]['gene_symbol']
```
Comment (mccalluc) [inline examples/differential-expression.md:9]: Any particular reason for `[10]`?

File: examples/differential-expression.md [hunk 7]
```diff
@@ -1,23 +1,26 @@
...
+>>> organs_with_gene = client.select_organs(where='gene', has=[gene_symbol], genomic_modality=modality, p_value=1.0)
+>>> organs_with_gene_details = organs_with_gene.get_list()[0:10]
+>>> organs_with_gene_details[0].keys()
+dict_keys(['grouping_name'])
```
Comment (mccalluc) [inline examples/differential-expression.md:13]: We've gone from `['gene_symbol', 'go_terms']` to `['grouping_name']`, which is a lot more generic: Is a `grouping_name` purely internal, or will it correspond to some vocabulary users will know?

Comment (mccalluc) [reply examples/differential-expression.md:13]: oh: looking at the code below, it looks like grouping name is just an organ? Will it always be organ?

File: examples/differential-expression.md [hunk 8]
```diff
@@ -1,23 +1,26 @@
...
-Find organs that differentially express the a gene at the 0.01 significance level
+Find genes differentially expressed by the organ at significance level 1.0:
+
 ```python
->>> gene_symbol = client.select_genes().get_list()[10]['gene_symbol']
->>> organs_with_gene = client.select_organs(where='gene', has=[gene_symbol], genomic_modality='atac', p_value=0.05)
->>> organs_with_gene_details = organs_with_gene.get_list()[0:10]
->>> organs_with_gene_details[0].keys()
-dict_keys(['grouping_name'])
+>>> organ_name = organs_with_gene_details[0]['grouping_name']
```
Comment (mccalluc) [inline examples/differential-expression.md:20]: Where are you getting organ names from? We already have a mess of organ terms, and it would be good not to add to the mess. Did you consider using terms from a controlled vocabulary?

File: examples/differential-expression.md [hunk 9]
```diff
@@ -1,23 +1,26 @@
...
+>>> organ_genes = client.select_genes(where='organ', has=[organ_name], genomic_modality='atac', p_value=1.0)
```
Comment (mccalluc) [inline examples/differential-expression.md:21]: `1.0` is not a typical p value.

File: examples/select_cells.md [hunk 10]
```diff
@@ -12,17 +12,14 @@
...
+>>> gene_symbol = client.select_genes(where="modality", has=["rna"]).get_list()[0]['gene_symbol']
 >>> cells_with_gene = client.select_cells(where='gene', has=[f'{gene_symbol} > 1'], genomic_modality='rna')
 >>> assert len(cells_with_gene) > 0
 
 >>> cells_with_gene_details_with_values = cells_with_gene.get_list(values_included=[gene_symbol])
 >>> cells_keys = cells_with_gene_details_with_values[0]['values'].keys()
 >>> assert list(cells_keys) == [gene_symbol]
 
->>> gene_symbol = client.select_genes().get_list()[10]['gene_symbol']
->>> cells_with_gene_atac = client.select_cells(where='gene', has=[gene_symbol], genomic_modality='atac')
```
Comment (mccalluc) [inline examples/select_cells.md:24]: What has changed here? If `has` now needs to be an inequality, that seems reasonable, but if it's something else, please explain.

File: examples/select_clusters.md [hunk 11]
```diff
@@ -11,8 +11,8 @@
 
 `client.select_clusters(where='gene', ...)`:
 ```python
->>> gene_symbol = client.select_genes().get_list()[0]['gene_symbol']
->>> clusters_with_gene_set = client.select_clusters(where='gene', has=[gene_symbol], genomic_modality='rna', p_value=0.05)
+>>> gene_symbol = client.select_genes(where="modality", has=["rna"]).get_list()[0]['gene_symbol']
+>>> clusters_with_gene_set = client.select_clusters(where='gene', has=[gene_symbol], genomic_modality='rna', p_value=1.0)
```
Comment (mccalluc) [inline examples/select_clusters.md:15]: Again, not a typical p value.

File: examples/select_datasets.md [hunk 12]
```diff
@@ -20,24 +20,25 @@
 
 `client.select_datasets(where='cluster', ...)`:
 ```python
->>> cluster_name = client.select_clusters().get_list()[30]['grouping_name']
```
Comment (mccalluc) [inline examples/select_datasets.md:23]: Do unconstrained searches no longer work? Is there an expected error behavior if a user attempts an unconstrained search?

File: examples/select_datasets.md [hunk 13]
```diff
@@ -20,24 +20,25 @@
...
+>>> gene_datasets = client.select_datasets(where='gene', has=[f'{gene_symbol} > 1'], genomic_modality='rna', min_cell_percentage=5.0)
 >>> assert len(gene_datasets) > 0
 
 ```
 
 `client.select_datasets(where='protein', ...)`:
 ```python
 >>> protein_name = client.select_proteins().get_list()[0]['protein_id']
->>> protein_datasets = client.select_datasets(where='protein', has=[f'{protein_name} > 5000'], min_cell_percentage=5.0)
+>>> protein_datasets = client.select_datasets(where='protein', has=[f'{protein_name} > 1000'], min_cell_percentage=5.0)
```
Comment (mccalluc) [inline examples/select_datasets.md:41]: Can you say anything about the semantics of this value, whether is `5000` or `1000`? Or is it arbitrary, and the pipelines now return smaller numbers?

File: examples/select_genes.md [hunk 14]
```diff
@@ -11,8 +11,11 @@
...
->>> organ_name = client.select_organs().get_list()[1]['grouping_name']
->>> organ_genes = client.select_genes(where='organ', has=[organ_name], genomic_modality='rna', p_value=0.05)
+>>> gene_symbol = client.select_genes(where="modality", has=["rna"]).get_list()[0]['gene_symbol']
+>>> organs_with_gene = client.select_organs(where='gene', has=[gene_symbol], genomic_modality='rna', p_value=1.0)
+>>> organs_with_gene_details = organs_with_gene.get_list()[0:10]
+>>> organ_name = organs_with_gene_details[0]['grouping_name']
+>>> organ_genes = client.select_genes(where='organ', has=[organ_name], genomic_modality='rna', p_value=1.0)
 
 >>> organ_genes_details = organ_genes.get_list()
 >>> organ_genes_details[0].keys()
```
Comment (mccalluc) [inline examples/select_genes.md:21]: Can you fill in the `where='cluster'` test below?

**Review by mccalluc — COMMENTED** [review]
Two big questions:
- Do unconstrained searches no longer work? If not, what is the expected error behavior?
- p-value of 1.0 seems strange to me, and would probably seem strange to users. Why this value?

Lots of smaller questions.

File: README.md (hunk 5 above)
Comment (SFD5311) [reply README.md:26]: The old query still works, but because no ordering of the genes in the database is enforced, it is not guaranteed to return a gene which can be used for a follow-up query where the genomic_modality is rna.  Constraining this search ensures we're not getting back a gene which isn't present in the RNA index and then trying to use it for an RNA based query.

File: examples/differential-expression.md (hunk 6 above)
Comment (SFD5311) [reply examples/differential-expression.md:9]: No, that looks like a typo that I will fix.  Thanks for catching.

File: examples/differential-expression.md (hunk 7 above)
Comment (SFD5311) [reply examples/differential-expression.md:13]: Behind the scenes, Datasets, Organs, Clusters, and soon Cell Types, are all subclasses of a class called Cell Grouping.  grouping_name refers generically to identifiers for any of these types of Cell Grouping.

File: examples/differential-expression.md (hunk 8 above)
Comment (SFD5311) [reply examples/differential-expression.md:20]: I'm getting two letter abbreviations from search-api, expanding them out using a mapping from a markdown file borrowed from another repo and resolving some of them (like Right Kidney and Left Kidney both to Kidney).  I've been trying to stay harmonized with other HuBMAP efforts on this, but I agree that finding a way to enforce that harmonization would be good.  I think there was some discussion a while back of a microservice which would resolve from two letter abbreviations to full organ names from a shared vocabulary.  Do you remember that/know if anything came of it?

File: examples/select_datasets.md (hunk 13 above)
Comment (SFD5311) [reply examples/select_datasets.md:41]: This change was to make the query hit one of the values we precomputed, mostly just to make the tests go faster.

File: examples/select_datasets.md (hunk 12 above)
Comment (SFD5311) [reply examples/select_datasets.md:23]: Not all clusters are guaranteed to be good input to cluster->dataset queries.  If a cluster isn't, the query should return, but the assert statement at the end of this block will fail.

File: examples/differential-expression.md (hunk 9 above)
Comment (SFD5311) [reply examples/differential-expression.md:21]: It is the value which gives us our highest probability of a non-empty query result here, regardless of index version.  If we want to make this a more sensible p value, then I'd need to either restructure the tests to not fail in response to empty results or more fully decouple the client tests from server behavior.

File: examples/select_genes.md (hunk 14 above)
Comment (SFD5311) [reply examples/select_genes.md:21]: Yes, I can do that.

**Comment by SFD5311** [comment]
It appears that the recent batch of changes to server code addressed the issue that was causing the intermittent 502s that we were seeing when running these tests.  I'm going to close that issue on the server code repo and open an issue for more comprehensive load testing to help catch similar issues sooner in the future.  The tests seem to have no problem completing now, so I think now would be a good time to merge this branch or to pull these changes and resume work on the notebook tests PR.  Once I'm back from vacation (4/24), I'm going to start work on another PR which will expose the Cells API's new cell type support via the client.  I think after that would be a good time to release a new version to PyPI.

File: examples/differential-expression.md (hunk 8 above)
Comment (mccalluc) [reply examples/differential-expression.md:20]: I've filed
- #73

File: examples/differential-expression.md (hunk 9 above)
Comment (mccalluc) [reply examples/differential-expression.md:21]: Filed
- #74

File: README.md (hunk 5 above)
Comment (mccalluc) [reply README.md:26]: This is the same issue as https://github.com/hubmapconsortium/portal-ui/issues/2584. I believe this is still easy for users to trip over. Put yourself in their shoes.

File: examples/differential-expression.md (hunk 6 above)
Comment (mccalluc) [reply examples/differential-expression.md:9]: Please fix typo.

File: examples/differential-expression.md (hunk 7 above)
Comment (mccalluc) [reply examples/differential-expression.md:13]: Please explain this in the documentation: If I'm confused in during the review, it's likely that any other user would also be confused, and they aren't going to dig through old PR comments.

File: examples/differential-expression.md (hunk 8 above)
Comment (mccalluc) [reply examples/differential-expression.md:20]: Again, anyone else reading the docs could have the same question.

File: examples/select_cells.md (hunk 10 above)
Comment (mccalluc) [reply examples/select_cells.md:24]: explanation?

File: examples/select_genes.md (hunk 14 above)
Comment (mccalluc) [reply examples/select_genes.md:21]: Doesn't seem to be done.

**Review by mccalluc — CHANGES_REQUESTED** [review]
Action items from previous review have not been addressed.
